src/Hamiltonians/excitations.jl

```diff
@@ -0,0 +1,252 @@
+        end
+        src_indices = getindex.(Ref(map), (f_hole, s_hole))
+        f_mode, s_mode = src_indices[1].mode, src_indices[2].mode
+        if mom_change ≥ s_mode - f_mode
+            mom_change += 1 # to avoid putting particles back into the holes
+        end
+    end
+    # For higher dimensions, replace mod1 here with some geometry.
+    src_modes = getproperty.(src_indices, :mode)
+    dst_modes = src_modes .+ (mom_change, -mom_change)
+    if fold
+        dst_modes = mod1.(dst_modes, M)
+    elseif !all(1 .≤ dst_modes .≤ M)
+        return add, 0.0, src_modes..., -mom_change
+    end
+    dst_indices = find_mode(add, dst_modes)
+    return excitation(add, dst_indices, src_indices)..., src_modes..., -mom_change
+end
+
+function momentum_transfer_excitation(add::FermiFS, chosen::Integer, map; fold=true)
+    return add, 0.0, 0, 0, 0
+end
+
+@inline function momentum_transfer_excitation(
+    add_a, add_b, chosen, map_a, map_b; fold=true
+)
+    M = num_modes(add_a)
+    M == num_modes(add_b) || throw(ArgumentError("Addresses must have the same number of modes"))
+
+    src_a, remainder = fldmod1(chosen, (M - 1) * length(map_b))
+    dst_a, src_b = fldmod1(remainder, length(map_b))
+
+    src_a_index = map_a[src_a]
+    src_b_index = map_b[src_b]
+    src_a_mode = src_a_index.mode
+
+    if dst_a ≥ src_a_mode
+        dst_a += 1 # to skip the src_a
+    end
+    mom_change = dst_a - src_a_mode # change in momentun
+    dst_a_index = find_mode(add_a, dst_a)
+    src_b_mode = src_b_index.mode
+    dst_b = src_b_mode - mom_change
+
+    # Additional info returned with result:
+    params = (src_a_mode, src_b_mode, -mom_change)
+
+    if fold
+        dst_a = mod1(dst_a, M)
+        dst_b = mod1(dst_b, M) # enforce periodic boundary condition
+    elseif !(0 < dst_a ≤ M) || !(0 < dst_b ≤ M)
+        return add_a, add_b, 0.0, params...
+    end
+
+    dst_b_index = find_mode(add_b, dst_b)
+
+    new_add_a, val_a = excitation(add_a, (dst_a_index,), (src_a_index,))
+    new_add_b, val_b = excitation(add_b, (dst_b_index,), (src_b_index,))
+
+    return new_add_a, new_add_b, val_a * val_b, params...
+end
+
+"""
+    momentum_transfer_diagonal(map)
+
+The diagonal part of [`momentum_transfer_excitation`](@ref).
+"""
+function momentum_transfer_diagonal(map::BoseOccupiedModeMap)
+    onproduct = 0
+    for i in 1:length(map)
+        occ_i = map[i].occnum
+        onproduct += occ_i * (occ_i - 1)
+        for j in 1:i-1
+            occ_j = map[j].occnum
+            onproduct += 4 * occ_i * occ_j
+        end
+    end
+    return float(onproduct)
+end
+function momentum_transfer_diagonal(
+    map_a::FermiOccupiedModeMap, map_b::FermiOccupiedModeMap
+)
+    onproduct = 0
+    n1 = length(map_a)
+    n2 = length(map_b)
+
+    return float(2 * n1 * n2)
+end
+function kinetic_energy(kes, map)
+    value = 0.0
+    for index in map
+        value += kes[index.mode] * index.occnum
+    end
+    return value
+end
```

Comment:
I'm wondering whether we should define this functionality to be the result of a `dot` product. Something like
```suggestion
@propagate_inbounds function LinearAlgebra.dot(kes, map::OccupiedModeMap)
    @boundscheck length(kes) == length(map) || throw(ArgumentError("mismatching arguments for dot"))
    value = zero(eltype(kes))
    for index in map
        @inbounds value += kes[index.mode] * index.occnum
    end
    return value
end
```
I'm not sure whether the `@inbounds` would speed things up, though.

Reply:
I like the idea. I have implemented it, and removed `kinetic_energy`. Note that the `@boundscheck` can't be done this way. It's fine for `kes` to be much larger than `map`.